lib/msal-browser/src/app/PublicClientApplication.ts

```diff
@@ -130,54 +130,62 @@ export class PublicClientApplication implements IPublicClientApplication {
+    private getRedirectResponseHash(): string {
         // Get current location hash from window or cache.
         const { location: { hash } } = window;
         const isResponseHash = UrlString.hashContainsKnownProperties(hash);
+        const cachedHash = this.browserStorage.getItem(this.browserStorage.generateCacheKey(TemporaryCacheKeys.URL_HASH), CacheSchemaType.TEMPORARY) as string;
+        this.browserStorage.removeItem(this.browserStorage.generateCacheKey(TemporaryCacheKeys.URL_HASH));
 
-        // Check for interaction type
-        if (isResponseHash) {
+        const responseHash = isResponseHash ? hash : cachedHash;
+        if (responseHash) {
             // Deserialize hash fragment response parameters.
-            const hashUrlString: UrlString = new UrlString(hash);
-            const serverParams: ServerAuthorizationCodeResponse = hashUrlString.getDeserializedHash<ServerAuthorizationCodeResponse>();
+            const serverParams: ServerAuthorizationCodeResponse = StringUtils.queryStringToObject<ServerAuthorizationCodeResponse>(responseHash);
             const requestStateObj: RequestStateObject = ProtocolUtils.parseRequestState(this.browserCrypto, serverParams.state);
             const platformStateObj: BrowserStateObject = BrowserProtocolUtils.parseBrowserRequestState(this.browserCrypto, requestStateObj.libraryState.platformState);
-            if (StringUtils.isEmpty(platformStateObj.interactionType) || platformStateObj.interactionType !== InteractionType.REDIRECT) {
+            if (platformStateObj.interactionType !== InteractionType.REDIRECT) {
                 return null;
+            } else {
+                BrowserUtils.clearHash();
+                return responseHash;
             }
+        }      
+
+        return null;
+    }
+
+    /**
+     * Checks if navigateToLoginRequestUrl is set, and:
+     * - if true, performs logic to cache and navigate
+     * - if false, handles hash string and parses response
+     */
+    private async handleRedirectResponse(): Promise<AuthenticationResult | null> {
+        const responseHash = this.getRedirectResponseHash();
+        if (!responseHash) {
+            return null;
         }
 
-        const cachedHash = this.browserStorage.getItem(this.browserStorage.generateCacheKey(TemporaryCacheKeys.URL_HASH), CacheSchemaType.TEMPORARY) as string;
         const loginRequestUrl = this.browserStorage.getItem(this.browserStorage.generateCacheKey(TemporaryCacheKeys.ORIGIN_URI), CacheSchemaType.TEMPORARY) as string;
         const loginRequestUrlNormalized = UrlString.removeHashFromUrl(loginRequestUrl || "");
         const currentUrlNormalized = UrlString.removeHashFromUrl(window.location.href);
 
-        if (loginRequestUrlNormalized === currentUrlNormalized && this.config.auth.navigateToLoginRequestUrl) {
+        if (loginRequestUrlNormalized === currentUrlNormalized || !this.config.auth.navigateToLoginRequestUrl) {
```

Comment:
You don't need the second check if you're going to check for truthiness inside the if. One reason I made these ifs top level is to make the logic easier to follow. I'd prefer to revisit this together.

Reply:
Reverted to be top level